### PR TITLE
[testplan] Add testplan for tunnel QoS remapping

### PR DESCRIPTION
**Pull request by bingwang-ms**
Signed-off-by: bingwang <wang.bing@microsoft.com>

### Description of PR

Summary:
This PR is to add detailed testplan for tunnel QoS remapping.
HLD [tunnel_dscp_remapping.md](https://github.com/Azure/SONiC/blob/master/doc/qos/tunnel_dscp_remapping.md)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012

### CODE REVIEW

File: docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+3. Randomly select a server (say ip `192.168.0.2`) facing interface on the selected ToR in step 1 (`upper_tor`), and toggle the mux status to `standby` in order to do packet encapsulation. server `192.168.0.2` would be active on `lower_tor`
+4. Startup `garp_service` on ptf to populate arp for servers
+
+### Test cases
+
+#### Test case 1 - Verify packets enter expected PG on lower_tor
+##### Test steps
+1. Generate `100` packets with `DSCP = 3/4`, `target_ip = 192.168.0.2`, `src_ip = 1.1.1.1`
+2. Send the `100` packets to `upper_tor` via a portchannel, and the encapped packets will be bounced to `lower_tor` via `T1`
+3. Verify the encapped packets is ingressed to the expected PG on `lower_tor` by sai_thrift api sai_thrift_read_pg_counters. (`DSCP 3 -> PG 2, DSCP 4 -> PG 6`)
```
Comment (lguohan) [inline docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md:66]: you should be able to get pg counter from sonic cli, not sure if you have to use syncd_rpc docerk.

File: docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+2. Randomly select a ToR for testing, say `upper_tor`, and the unselected ToR would be `lower_tor`
+3. Randomly select a server (say ip `192.168.0.2`) facing interface on the selected ToR in step 1 (`upper_tor`), and toggle the mux status to `standby` in order to do packet encapsulation. server `192.168.0.2` would be active on `lower_tor`
+4. Startup `garp_service` on ptf to populate arp for servers
+
+### Test cases
+
+#### Test case 1 - Verify packets enter expected PG on lower_tor
+##### Test steps
+1. Generate `100` packets with `DSCP = 3/4`, `target_ip = 192.168.0.2`, `src_ip = 1.1.1.1`
+2. Send the `100` packets to `upper_tor` via a portchannel, and the encapped packets will be bounced to `lower_tor` via `T1`
```
Comment (neethajohn) [inline docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md:65]: For our ptf setup, we need to send encapped packet (for eg. outer dscp: 2, inner dscp: 3) from T1 to active tor and verify that it is mapped to the correct pg (2)
As in other cases, please include dscps that map to other TC's as well and verify all others except lossless map to pg 0

File: docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md (hunk 1 above)
Comment (neethajohn) [reply docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md:66]: sonic cli only has pg dropped counters. we will still need to use syncd rpc docker for this test

File: docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+3. Verify the encapped packets is ingressed to the expected PG on `lower_tor` by sai_thrift api sai_thrift_read_pg_counters. (`DSCP 3 -> PG 2, DSCP 4 -> PG 6`)
+
+#### Test case 2 - Verify packets egressed to server at expected queue
+##### Test steps
+1. Generate `100` packets with `DSCP = 3/4`, `target_ip = 192.168.0.2`, `src_ip = 1.1.1.1`
+2. Send the `100` packets to `upper_tor` via a portchannel
+3. Verify the packets are bounced back to `lower_tor`, and the outlayer DSCP value is as expected (3->2, 4->6)
```
Comment (neethajohn) [inline docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md:72]: In the ptf infrastructure, we should sent encapped packets to the active ToR and verify this behavior. Please verify other dscp values as well

File: docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+4. Verify the decapped packets is egressed to server at expected queue with CLI `show queue counter`. The packet counter for expected queue is supposed to be larger or equal to `100`. (`DSCP 3 -> Queue 3, DSCP 4 -> Queue 4`)
+
+#### Test case 3 - Verify PFC frame generation at expected queue
+##### Test steps
+1. Disable egress for T1 facing port on `lower_tor` with sai_thrift api `sai_thrift_port_tx_disable`
+2. Generate packet with `DSCP = 3/4`, `target_ip = 192.168.0.2`, `src_ip = 1.1.1.1`. 
+3. Send the packet to `upper_tor` via a portchannel, and the traffic is bounced back to T1. The number of transmitted packets is determined by the lossless profile to ensure the pg is filled.
+4. Send the packet one more time, and verify `lower_tor` will generate PFC pause frame on expected queue with sai_thrift api `sai_thrift_read_port_counters` (`DSCP 3 -> PFC 2,  DSCP 4 -> PFC 6`).
```
Comment (neethajohn) [inline docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md:80]: Please look at the xoff test below. We need to do the same here sending encapped packets to the active ToR testing both outer dscp 2 and 6. basically determine the number of packets to fill up the shared space completely and ensure that PFC is triggered on the correct priority. Then send few more packets exceeding headroom and verify that the correct PG drop counters are incremented.
https://github.com/Azure/sonic-mgmt/blob/master/tests/saitests/sai_qos_tests.py#L649

File: docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+
```
Comment (neethajohn) [inline docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md:82]: We can also have a testcase where we send encapped (outer dscp: 2) and regular packets (dscp: 3) to the active ToR. Fill up the shared space and verify that pfc is triggered for both priority 2 and 3.

File: docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+#### Test case 2 - Verify packets egressed to server at expected queue
+##### Test steps
+1. Generate `100` packets with `DSCP = 3/4`, `target_ip = 192.168.0.2`, `src_ip = 1.1.1.1`
+2. Send the `100` packets to `upper_tor` via a portchannel
+3. Verify the packets are bounced back to `lower_tor`, and the outlayer DSCP value is as expected (3->2, 4->6)
+4. Verify the decapped packets is egressed to server at expected queue with CLI `show queue counter`. The packet counter for expected queue is supposed to be larger or equal to `100`. (`DSCP 3 -> Queue 3, DSCP 4 -> Queue 4`)
+
+#### Test case 3 - Verify PFC frame generation at expected queue
+##### Test steps
+1. Disable egress for T1 facing port on `lower_tor` with sai_thrift api `sai_thrift_port_tx_disable`
```
[the remark this answers is not in the document]
Comment (neethajohn) [reply docs/testplan/QoS-remapping-for-Tunnel-traffic-test-plan.md:77]: we need to have PG stat counters also exposed to fully get rid of syncd rpc

File: tests/vtestbed.csv [hunk 7]
```diff
@@ -1 +0,0 @@
-../ansible/vtestbed.csv
```
Comment (neethajohn) [inline tests/vtestbed.csv:1]: remove this file

Comment (bingwang-ms) [reply tests/vtestbed.csv]: Oh, remove by mistake. Fixed now. Thanks

**Review by neethajohn — APPROVED** [review]

**Pull request merged**